Add Heroku limitations section. Resolves #701.

jesseshieh opened the pull request:
As discussed in issue #701, many beginners are surprised to find out later on that Heroku has some serious limitations for Elixir apps. This pull request adds a limitations section to the Heroku guide to help them avoid being bitten down the road.

ericmj commented:
Thanks @jesseshieh. I think limitations works better at the end of the guide. I also think it's wrong to include a specific alternative to Heroku, since you could also list Dokku, Zeit, AWS, Google Cloud, Linode, DigitalOcean etc.

josevalim commented on docs/deployment/heroku.md:
```diff
@@ -8,6 +8,27 @@ The only thing we'll need for this guide is a working Phoenix application. For t
+- [Remote shells](https://hexdocs.pm/iex/IEx.html#module-remote-shells) and remote observer are not possible. 
+    - Heroku does not allow SSH access to your dynos so you can not inspect, debug, or trace your production nodes using things like [the built-in Observer](https://elixir-lang.org/getting-started/mix-otp/supervisor-and-application.html#observer).
+
+- Hot upgrades are not possible. 
```
Can we please remove this item? Most people should not be using hot upgrades, regardless of Heroku or not. :)

jesseshieh replied:
Fair point! I'll remove it.

josevalim commented on docs/deployment/heroku.md:
```diff
@@ -8,6 +8,27 @@ The only thing we'll need for this guide is a working Phoenix application. For t
 
+### Limitations
+
+Heroku is a great platform overall, but for Elixir apps, it comes with a number of limitations you should be aware of. 
```
Thank you @jesseshieh!

Listing those limitations are important but they sound like a blocker for the use of Phoenix on Heroku. We need to make it clear that they come to mind when you need to push the envelope. What about something like:

> Heroku is a great platform and Elixir performs well on it. However, you may run into limitations if you plan to leverage advanced features provided by Elixir and Phoenix, such as:

Then for the closing paragraph:

> If you are just getting started or you don't expect to use the features above, Heroku should be enough for your needs. For instance, if you are migrating an existing application running on Heroku to Phoenix, keeping a similar set of features, Elixir will perform just as well or even better then your current stack.

jesseshieh replied:
Oh yes, absolutely. I'll make those changes. I didn't mean for it to sound like a blocker. 

My goal in building Gigalixir was actually to help increase Elixir adoption since I fell in love with the language and ecosystem. I noticed that deployment seemed to be the most common pain point for people new to Elixir so I thought making deployment easier would make it more likely for people to choose Elixir. If you agree with that, do you think including a reference to Gigalixir somewhere in the docs would be mutually beneficial to Elixir, Gigalixir, and the developers?

josevalim replied:
It is great that are starting to see solutions that leverages all Elixir has to offer :D I am afraid though if you just drop Gigaelixir, it may do more harm than good, because how will the users even get started?

So what if, when suggesting alternatives, we link to a Gigaelixir guide? Could that be hosted on Gigaelixir? Let's also mention the distillery guide, provided by Phoenix, as an alternative.

jesseshieh replied:
Yes! There is a [Gigalixir guide](http://gigalixir.readthedocs.io/en/latest/main.html#quick-start) hosted on Gigalixir I can link to. I'll also link to the distillery guide.

jesseshieh replied:
Hm, it doesn't look like there is a distillery guide yet, but there is an exrm guide. I'll take a stab at writing a distillery guide for Phoenix if you like.

josevalim commented:
@chrismccord I am not sure if this should go at the bottom or at the top of the guide. Your call.

jesseshieh commented:
@josevalim pull request updated following your comments.

jesseshieh commented:
@josevalim, just following up on this pull request. I added the intro and closing paragraph, removed hot upgrades, changed the gigalixir link to the tutorial, and added distillery.

josevalim commented:
Thanks @jesseshieh! We should ping @Gazler and @chrismccord for merging it - as I don't really maintain the guides. :)

Gazler commented:
I think it is a good idea to point out the limitations of Heroku, as these are certainly things that come up. I am not sure that we should be providing a single alternative. I understand from https://elixirforum.com/t/gigalixir-platform-as-a-service-designed-just-for-elixir-phoenix/4946 that Gigalixir is your product and I think it is a great idea, but I am not sure it has its place in this section of the guides. You can see https://github.com/phoenixframework/phoenix/pull/2220#issuecomment-297468208 for more information.

After 1.3 is out, we should discuss a plan for moving guides for 3rd party services out of the main guides and simply link to them. This means that we aren't choosing to promote Heroku, and all third party services will get equal representation. This extends beyond PaaS services btw. For example, we have the same problem in the guides currently with email libraries.

With that in mind, I'd propose we remove the final section for now and open a new issue for how we plan to support 3rd party guides.

josevalim commented:
@Gazler I think pointing out all of those problems and not providing a solution can be confusing to the readers. So if we don't want to provide two examples (today we provide Distillery / Gigaelixir) then I would try to provide more examples of platforms without those downsides.

josevalim commented on docs/deployment/heroku.md:
```diff
@@ -8,6 +8,26 @@ The only thing we'll need for this guide is a working Phoenix application. For t
+
+If you are just getting started or you don't expect to use the features above, Heroku should be enough for your needs. For instance, if you are migrating an existing application running on Heroku to Phoenix, keeping a similar set of features, Elixir will perform just as well or even better then your current stack.
+ 
+If you want a platform-as-a-service without these limitations, try [Gigalixir](http://gigalixir.readthedocs.io/). If you don't want a PaaS, try [Distillery](https://github.com/bitwalker/distillery).
```
Instead of "If you don't want a PaaS", what if we say "If you would rather deploy to a cloud platform, such as EC2, Google Cloud, etc, consider [Distillery](https://github.com/bitwalker/distillery)."

Not sure if cloud platform is the best term here but I think you get the idea?

jesseshieh replied:
Done. I couldn't think of a better term than "cloud platform" so I just stuck with what you wrote.

jesseshieh commented:
@Gazler @josevalim any resolution on this? I'm happy to make whatever changes you guys want, but I'm getting mixed messages.

josevalim commented:
@jesseshieh it is still :+1: from me but @Gazler and @chrismccord are the ones who decide.

jesseshieh commented:
@Gazler @chrismccord lemme know what you want me to do and I'll fix it up. For what it's worth, gigalixir is definitely not motivated by profit so this isn't just some company trying to get free advertising. I'm honestly just trying to help out the community and increase Elixir adoption. I'm hoping that mentioning gigalixir will help solve a pain point that may be holding some people back from using Elixir.

Gazler commented:
@jessieshieh that makes sense. Thanks for your patience with this issue. I think what we should do is the include gigalixir as an alternative for now (along with distillery), we can always move it out to a different section when such a section exists. I'll still create a separate issue for deployment guide links.

jesseshieh commented:
@gazler, in that case, I think this pull request is ready to merge!

Gazler commented:
Thanks <3. Sorry it took so long to merge.
